File: benchmarks/Benchmarks/MNIST.cs
// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
using System;
using System.IO;
using System.Linq;
using System.Collections.Generic;
using System.Diagnostics;

using TorchSharp;
using TorchSharp.torchvision;

using TorchSharp.Examples;
using TorchSharp.Examples.Utils;

using static TorchSharp.torch;

using static TorchSharp.torch.nn;
using static TorchSharp.torch.nn.functional;

namespace CSharpExamples
{
    /// <summary>
    /// Simple MNIST Convolutional model.
    /// </summary>
    /// <remarks>
    /// There are at least two interesting data sets to use with this example:
    /// 
    /// 1. The classic MNIST set of 60000 images of handwritten digits.
    ///
    ///     It is available at: http://yann.lecun.com/exdb/mnist/
    ///     
    /// 2. The 'fashion-mnist' data set, which has the exact same file names and format as MNIST, but is a harder
    ///    data set to train on. It's just as large as MNIST, and has the same 60/10 split of training and test
    ///    data.
    ///    It is available at: https://github.com/zalandoresearch/fashion-mnist/tree/master/data/fashion
    ///
    /// In each case, there are four .gz files to download. Place them in a folder and then point the '_dataLocation'
    /// constant below at the folder location.
    /// </remarks>
    public class MNIST
    {
        private static int _epochs = 4;
        private static int _trainBatchSize = 64;

        private readonly static int miniBatchMultiplier = 100;

        internal static void Run(int epochs, string benchmarks, int batchsize)
        {
            _epochs = epochs;
            _trainBatchSize = batchsize;

            var device = cuda.is_available() ? CUDA : CPU;

            var data = Enumerable.Range(0, miniBatchMultiplier).Select(i => torch.rand(_trainBatchSize, 1, 28, 28, device: device)).ToList();
            var labels = Enumerable.Range(0, miniBatchMultiplier).Select(i => torch.randint(0, 10, _trainBatchSize, dtype: int64, device: device)).ToList();

            if (benchmarks.Contains("FWDU")) ForwardPassWithUsings(data, device);
            if (benchmarks.Contains("FWDNU")) ForwardPassNoUsings(data, device);

            if (benchmarks.Contains("BPNU")) BackpropNoUsings(data, labels, device);
            if (benchmarks.Contains("BPU")) BackpropWithUsings(data, labels, device);
        }

        private static void ForwardPassWithUsings(IList<Tensor> data, Device device)
        {
            Console.WriteLine();
            Console.WriteLine($"\tRunning MNIST ForwardPassWithUsings with random data on {device.type.ToString()} for {_epochs}*{miniBatchMultiplier} iterations with a batch size of {_trainBatchSize}.");

            var model = new TorchSharp.Examples.MNIST.Model(device);

            // Warm-up

            using (var output = model.forward(data[0])) { }

            Console.WriteLine($"\tStart timing...");
            Stopwatch totalTime = new Stopwatch();
            totalTime.Start();

            for (int i = 0; i < _epochs; i++)
            {
                for (var batch = 0; batch < data.Count(); batch++)
                {
                    using (var output = model.forward(data[batch])) { }
                }
            }
            totalTime.Stop();
            Console.WriteLine($"\tElapsed time: {totalTime.Elapsed.TotalSeconds:F1} s.");
        }

        private static void ForwardPassNoUsings(IList<Tensor> data, Device device)
        {
            Console.WriteLine();
            Console.WriteLine($"\tRunning MNIST ForwardPassNoUsings with random data on {device.type.ToString()} for {_epochs}*{miniBatchMultiplier} iterations with a batch size of {_trainBatchSize}.");

            var model = new TorchSharp.Examples.MNIST.Model(device);

            // Warm-up

            using (var output = model.forwardNoUsings(data[0])) { }
            GC.Collect();

            Console.WriteLine($"\tStart timing...");
            Stopwatch totalTime = new Stopwatch();
            totalTime.Start();

            for (int i = 0; i < _epochs; i++)
            {
                for (var batch = 0; batch < data.Count(); batch++)
                {
                    var output = model.forwardNoUsings(data[batch]);
                    GC.Collect();
                }
            }

            totalTime.Stop();
            Console.WriteLine($"\tElapsed time: {totalTime.Elapsed.TotalSeconds:F1} s.");
        }

        private static void BackpropWithUsings(IList<Tensor> data, IList<Tensor> labels, Device device)
        {
            Console.WriteLine();
            Console.WriteLine($"\tRunning MNIST BackpropWithUsings with random data on {device.type.ToString()} for {_epochs}*{miniBatchMultiplier} iterations with a batch size of {_trainBatchSize}.");

            var model = new TorchSharp.Examples.MNIST.Model(device);

            // Warm-up

            using (var output = model.forward(data[0])) { }

            model.Train();

            var loss = nll_loss(reduction: Reduction.Mean);
            var optimizer = optim.Adam(model.parameters());

            model.Train();

            Console.WriteLine($"\tStart timing...");
            Stopwatch totalTime = new Stopwatch();
            totalTime.Start();

            for (int i = 0; i < _epochs; i++)
            {
                for (var batch = 0; batch < data.Count(); batch++)
                {
                    optimizer.zero_grad();

                    using (var prediction = model.forward(data[batch]))
                    using (var output = loss(prediction, labels[batch]))
                    {
                        output.backward();
                        optimizer.step();
                    }
                }
            }

            totalTime.Stop();
            Console.WriteLine($"\tElapsed time: {totalTime.Elapsed.TotalSeconds:F1} s.");
        }

        private static void BackpropNoUsings(IList<Tensor> data, IList<Tensor> labels, Device device)
        {
            Console.WriteLine();
            Console.WriteLine($"\tRunning MNIST BackpropNoUsings with random data on {device.type.ToString()} for {_epochs}*{miniBatchMultiplier} iterations with a batch size of {_trainBatchSize}.");
            Console.WriteLine($"\tCreating the model...");

            var model = new TorchSharp.Examples.MNIST.Model(device);

            // Warm-up

            using (var output = model.forwardNoUsings(data[0])) { }

            var loss = nll_loss(reduction: Reduction.Mean);
            var optimizer = optim.Adam(model.parameters());

            model.Train();

            Console.WriteLine($"\tStart timing...");
            Stopwatch totalTime = new Stopwatch();
            totalTime.Start();

            for (int i = 0; i < _epochs; i++)
            {
                for (var batch = 0; batch < data.Count(); batch++)
                {
                    optimizer.zero_grad();

                    var prediction = model.forwardNoUsings(data[batch]);
                    var output = loss(prediction, labels[batch]);

                    output.backward();
                    optimizer.step();

                    GC.Collect();
                }
            }

            totalTime.Stop();
            Console.WriteLine($"\tElapsed time: {totalTime.Elapsed.TotalSeconds:F1} s.");
        }
    }
}

File: benchmarks/python/mnist.py
from __future__ import print_function
import argparse
import time
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR

# import exportsd

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.seq = nn.Sequential(
          nn.Conv2d(1, 32, 3, 1),
          nn.ReLU(),
          nn.Conv2d(32, 64, 3, 1),
          nn.ReLU(),
          nn.MaxPool2d(2),
          nn.Dropout(0.25),
          nn.Flatten(),
          nn.Linear(9216, 128),
          nn.ReLU(),
          nn.Dropout(0.5),
          nn.Linear(128, 10)
        )

    def forward(self, x):
        x = self.seq(x)
        output = F.log_softmax(x, dim=1)
        return output


def forwardPassOnly(model, epochs, data, device):

    print("\tRunning MNIST forwardPassOnly with random data for", epochs,"*", len(data), "iterations with a batch size of", data[0].shape[0], " running on", device)

    # Warm-up

    _ = model(data[0])

    start_time = time.time()

    for i in range(epochs):
        for d in data:
            
            model(d)

    elapsed = time.time() - start_time

    print('Elapsed time: {:5.2f}s'.format(elapsed))

def backpropagation(model, epochs, data, labels, device):

    print("\tRunning MNIST forwardPassOnly with random data for", epochs,"*", len(data), "iterations with a batch size of", data[0].shape[0], " running on", device)

    # Warm-up

    _ = model(data[0])

    optimizer = optim.Adam(model.parameters(), lr=0.001)

    model.train()
    
    start_time = time.time()

    for i in range(epochs):
        for di in range(len(data)):

            optimizer.zero_grad()

            prediction = model(data[di])
            output = F.nll_loss(prediction, labels[di])

            output.backward()

            optimizer.step()

    elapsed = time.time() - start_time

    print('Elapsed time: {:5.2f}s'.format(elapsed))

def main():
    # Training settings
    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--batches', type=int, default=100, metavar='N',
                        help='the number of simulated mini-batched (default: 100)')
    parser.add_argument('--batch-size', type=int, default=512, metavar='N',
                        help='input batch size for training (default: 512)')
    parser.add_argument('--epochs', type=int, default=10, metavar='N',
                        help='number of epochs to train (default: 10)')
    parser.add_argument('--no-cuda', action='store_true', default=False,
                        help='disables CUDA training')
    parser.add_argument('--benchmark', type=str, default="FWD|BPRP",
                        help='disables CUDA training')
    args = parser.parse_args()
    use_cuda = not args.no_cuda and torch.cuda.is_available()

    device = torch.device("cuda" if use_cuda else "cpu")

    train_kwargs = {'batch_size': args.batch_size}
    if use_cuda:
        cuda_kwargs = {'num_workers': 1,
                       'pin_memory': True,
                       'shuffle': True}
        train_kwargs.update(cuda_kwargs)

    data = [torch.rand(args.batch_size, 1, 28, 28).to(device) for _ in range(args.batches)]
    labels = [torch.randint(0, 10, (args.batch_size,), dtype=torch.int64).to(device) for _ in range(args.batches)]

    model = Net().to(device)

    benchmarks = args.benchmark.split('|')

    for b in benchmarks:
        if b == "FWD": 
            forwardPassOnly(model, args.epochs, data, device)
        elif b == "BPRP":
            backpropagation(model, args.epochs, data, labels, device)
        

if __name__ == '__main__':
    main()
